flash-lso/src/amf3/read.rs

```diff
@@ -464,145 +473,225 @@ impl AMF3Decoder {
-            }
+                // Must parse `length` elements
+                let (i, el) =
+                    many_m_n(length_usize, length_usize, |i| this.parse_single_element(i))(i)?;
+
+                let elements_len = elements.len() as u32;
 
-            // Must parse `length` elements
-            let (i, el) =
-                many_m_n(length_usize, length_usize, |i| this.parse_single_element(i))(i)?;
+                let id = if let Value::ECMAArray(id, _, _, _) =
+                    this.object_reference_table.get(ofi).unwrap().as_ref()
+                {
+                    id
+                } else {
+                    unreachable!("{:?}", this.object_reference_table.get(ofi))
+                };
 
-            let elements_len = elements.len() as u32;
-            Ok((i, Value::ECMAArray(el, elements, elements_len)))
-        })
+                Ok((i, Value::ECMAArray(*id, el, elements, elements_len)))
+            },
+        )
     }
 
     fn parse_element_dict<'a>(&mut self, i: &'a [u8]) -> AMFResult<'a, Rc<Value>> {
-        self.parse_reference_or_val(i, |this, i, len| {
-            //TODO: implications of this
-            let (i, weak_keys) = be_u8(i)?;
+        self.parse_reference_or_val(
+            i,
+            |this| {
+                this.object_id += 1;
+                Value::Dictionary(ObjectId(this.object_id), Vec::new(), false)
+            },
+            |this, i, len, ofi| {
+                //TODO: implications of this
+                let (i, weak_keys) = be_u8(i)?;
 
-            // There must be at least `len * 2` bytes (due to (key,val) pairs) to read this, this prevents OOM errors with v.large dicts
-            if i.len() < len * 2 {
-                return Err(Err::Error(make_error(i, ErrorKind::TooLarge)));
-            }
+                // There must be at least `len * 2` bytes (due to (key,val) pairs) to read this, this prevents OOM errors with v.large dicts
+                if i.len() < len * 2 {
+                    return Err(Err::Error(make_error(i, ErrorKind::TooLarge)));
+                }
 
-            let (i, pairs) = many_m_n(len * 2, len * 2, |i| this.parse_single_element(i))(i)?;
+                let (i, pairs) = many_m_n(len * 2, len * 2, |i| this.parse_single_element(i))(i)?;
 
-            let pairs = pairs
-                .chunks_exact(2)
-                .map(|chunk| (chunk[0].clone(), chunk[1].clone()))
-                .collect::<Vec<_>>();
+                let pairs = pairs
+                    .chunks_exact(2)
+                    .map(|chunk| (chunk[0].clone(), chunk[1].clone()))
+                    .collect::<Vec<_>>();
 
-            Ok((i, Value::Dictionary(pairs, weak_keys == 1)))
-        })
+                let id = if let Value::Dictionary(id, _, _) =
+                    this.object_reference_table.get(ofi).unwrap().as_ref()
+                {
+                    id
+                } else {
+                    unreachable!()
```

Comment:
nit: A message would be appreciated here, for example like at `Value::ECMAArray`.

Reply:
Done